File: 4일차/길웅/가장먼노드.py
# 프로그래머스 '가장 먼 노드'
# https://programmers.co.kr/learn/courses/30/lessons/49189

from collections import deque

def solution(n, edge):
    node_dict = dict()

    for e1, e2 in edge:
        if e1 not in node_dict:
            node_dict[e1] = list()
        
        if e2 not in node_dict:
            node_dict[e2] = list()
            
        node_dict[e1].append(e2)
        node_dict[e2].append(e1)
    
    que = deque()
    visits = [False] * (n + 1)

    node = 1
    level = 0
    que.append((node, level))
    visits[1] = True
    
    max_level = 0
    answer = 0

    while que:
        node, level = que.popleft()
        
        if max_level < level:
            max_level = level
            answer = 1
        else:
            answer += 1

        next_nodes = node_dict.get(node, None)

        if next_nodes == None:
            continue

        for next_node in next_nodes:
            if visits[next_node]:
                continue

            que.append((next_node, level + 1))
            visits[next_node] = True

    return answer

File: 4일차/현동/가장먼노드.js
// https://programmers.co.kr/learn/courses/30/lessons/49189

function solution(n, edge) {
  let q = [1];
  let visited = Array(n + 1).fill(0);
  let Map = Array.from({ length: n + 1 }, () => []);
  visited[1] = 1;

  edge.forEach((e) => {
    Map[e[0]].push(e[1]);
    Map[e[1]].push(e[0]);
  });

  let max = 0;
  while (q.length !== 0) {
    let node = q.shift();
    Map[node].forEach((next) => {
      if (!visited[next]) {
        q.push(next);
        visited[next] = visited[node] + 1;
        max = visited[node] + 1;
      }
    });
  }

  return visited.filter((v) => v === max).length;
}
